File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/translate/DataTranslation.java
package uk.co.cerihughes.mgm.translate;

import org.openapitools.model.AlbumApiModel;
import org.openapitools.model.EventApiModel;
import org.openapitools.model.LocationApiModel;
import org.openapitools.model.PlaylistApiModel;
import uk.co.cerihughes.mgm.model.interim.InterimAlbum;
import uk.co.cerihughes.mgm.model.interim.InterimEvent;
import uk.co.cerihughes.mgm.model.interim.InterimPlaylist;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class DataTranslation {
    private final List<AlbumTranslation> albumTranslations = new ArrayList<>();
    private final List<PlaylistTranslation> playlistTranslations = new ArrayList<>();

    public void addAlbumTranslation(AlbumTranslation translation) {
        albumTranslations.add(translation);
    }

    public void addPlaylistTranslation(PlaylistTranslation tranlsation) {
        playlistTranslations.add(tranlsation);
    }

    public List<EventApiModel> translate(List<InterimEvent> interimEvents) {
        Stream.concat(albumTranslations.stream(), playlistTranslations.stream())
                .forEach(e -> e.preprocess(interimEvents));

        return interimEvents.stream()
                .map(e -> translate(e))
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    private EventApiModel translate(InterimEvent interimEvent) {
        final AlbumApiModel classicAlbum = translate(interimEvent.getClassicAlbum());
        final AlbumApiModel newAlbum = translate(interimEvent.getNewAlbum());
        final PlaylistApiModel playlist = translate(interimEvent.getPlaylist());

        EventApiModel model = new EventApiModel();
        model.setNumber(interimEvent.getNumber());
        model.setDate(interimEvent.getDate());
        model.setClassicAlbum(classicAlbum);
        model.setNewAlbum(newAlbum);
        model.setLocation(translateLocation());
        model.setPlaylist(playlist);
        return model;
    }

    private AlbumApiModel translate(InterimAlbum interimAlbum) {
        if (interimAlbum == null) {
            return null;
        }

        for (AlbumTranslation albumTranslation : albumTranslations) {
            AlbumApiModel outputAlbum = albumTranslation.translate(interimAlbum);
            if (outputAlbum != null) {
                return outputAlbum;
            }
        }

        return null;
    }

    private PlaylistApiModel translate(InterimPlaylist interimPlaylist) {
        if (interimPlaylist == null) {
            return null;
        }

        for (PlaylistTranslation playlistTranslation : playlistTranslations) {
            PlaylistApiModel outputPlaylist = playlistTranslation.translate(interimPlaylist);
            if (outputPlaylist != null) {
                return outputPlaylist;
            }
        }

        return null;
    }

    private LocationApiModel translateLocation() {
        final String name = "Crafty Devil's Cellar";
        final double latitude = 51.48227690;
        final double longitude = -3.20186570;

        LocationApiModel model = new LocationApiModel();
        model.setName(name);
        model.setLatitude(latitude);
        model.setLongitude(longitude);
        return model;
    }
}

File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/translate/DataTranslation.kt
package uk.co.cerihughes.mgm.translate

import uk.co.cerihughes.mgm.model.interim.InterimAlbum
import uk.co.cerihughes.mgm.model.interim.InterimEvent
import uk.co.cerihughes.mgm.model.interim.InterimLocation
import uk.co.cerihughes.mgm.model.interim.InterimPlaylist
import uk.co.cerihughes.mgm.model.output.AlbumApiModel
import uk.co.cerihughes.mgm.model.output.EventApiModel
import uk.co.cerihughes.mgm.model.output.LocationApiModel
import uk.co.cerihughes.mgm.model.output.PlaylistApiModel

class DataTranslation {
    companion object {
        private var chapterArtsCenter = createLocation("Chapter Arts Center", 51.4829773, -3.2056947)
        private var tenFeetTall = createLocation("10 Feet Tall", 51.4805194, -3.1810703)
        private var craftyDevilsCellar = createLocation("Crafty Devil's Cellar", 51.48227690, -3.20186570)
        private var stayAtHome = createLocation("#StayAtHome", 0.0, 0.0)

        private val locationMap = mapOf(
            Pair("CHAPTER", chapterArtsCenter),
            Pair("TEN_FEET", tenFeetTall),
            Pair("CRAFTY", craftyDevilsCellar),
            Pair("ZOOM", stayAtHome)
        )

        private fun createLocation(name: String, latitude: Double, longitude: Double): LocationApiModel {
            val model = LocationApiModel()
            model.name = name
            model.latitude = latitude
            model.longitude = longitude
            return model
        }
    }

    private val albumTranslations: MutableList<AlbumTranslation> = ArrayList()
    private val playlistTranslations: MutableList<PlaylistTranslation> = ArrayList()
    fun addAlbumTranslation(translation: AlbumTranslation) {
        albumTranslations.add(translation)
    }

    fun addPlaylistTranslation(tranlsation: PlaylistTranslation) {
        playlistTranslations.add(tranlsation)
    }

    fun translate(interimEvents: List<InterimEvent>): List<EventApiModel> {
        (albumTranslations + playlistTranslations).forEach { it.preprocess(interimEvents) }
        return interimEvents
            .mapNotNull { translate(it) }
    }

    private fun translate(interimEvent: InterimEvent): EventApiModel {
        val model = EventApiModel()
        model.number = interimEvent.number
        model.date = interimEvent.date
        model.classicAlbum = translate(interimEvent.classicAlbum)
        model.newAlbum = translate(interimEvent.newAlbum)
        model.location = translate(interimEvent.location)
        model.playlist = interimEvent.playlist?.let { translate(it) }
        return model
    }

    private fun translate(interimAlbum: InterimAlbum): AlbumApiModel? {
        return albumTranslations
            .mapNotNull { it.translate(interimAlbum) }
            .firstOrNull()
    }

    private fun translate(interimPlaylist: InterimPlaylist): PlaylistApiModel? {
        return playlistTranslations
            .mapNotNull { it.translate(interimPlaylist) }
            .firstOrNull()
    }

    private fun translate(interimLocation: InterimLocation?): LocationApiModel {
        return interimLocation?.locationData?.let {
            locationMap.get(it)
        } ?: craftyDevilsCellar
    }
}
